pkg/ccl/importccl/read_import_pgdump.go

```diff
@@ -567,22 +571,44 @@ func (m *pgDumpReader) readFile(
 			if ok && conv == nil {
 				return errors.Errorf("missing schema info for requested table %q", name)
 			}
+			expectedColLen := len(i.Columns)
+			if expectedColLen == 0 {
+				// Case where the targeted columns are not specified in the PGDUMP file, but in
+				// the command "IMPORT INTO table (targetCols) PGDUMP DATA (filename)"
+				expectedColLen = len(conv.VisibleCols)
+			}
 			values, ok := i.Rows.Select.(*tree.ValuesClause)
 			if !ok {
 				return errors.Errorf("unsupported: %s", i.Rows.Select)
 			}
 			inserts++
 			startingCount := count
+			colLookup := make(map[string]int, len(conv.VisibleCols))
+			for j, col := range conv.VisibleCols {
+				colLookup[col.Name] = j
+			}
 			for _, tuple := range values.Rows {
 				count++
 				if count <= resumePos {
 					continue
 				}
-				if expected, got := len(conv.VisibleCols), len(tuple); expected != got {
-					return errors.Errorf("expected %d values, got %d: %v", expected, got, tuple)
+				if got := len(tuple); expectedColLen != got {
+					return errors.Errorf("expected %d values, got %d: %v", expectedColLen, got, tuple)
 				}
-				for i, expr := range tuple {
-					typed, err := expr.TypeCheck(ctx, &semaCtx, conv.VisibleColTypes[i])
+				conv.IsTargetCol = make(map[int]struct{})
+				for j, expr := range tuple {
+					var ind int
+					if len(i.Columns) == 0 {
+						ind = j
+					} else {
+						colName := i.Columns[j].String()
```

Comment:
Ok so what we *actually* need is just the type of the column, and to produce an error if a name is used that doesn't exist. So maybe something like:

```
if len(i.Columns) == 0 {
  typ = conv.VisibleColTypes[j]
} else {
  colName := i.Columns[j].String()
  typ, ok = conv.SomeNameLookup[colName]
  if !ok { ... }
}
```

And then set `SomeNameLookup` (renamed clearly) one time during init. It is possible some variable like this already exists, I haven't looked at what `conv` is. Does that work? If so it would remove creating 2 maps per insert row, which is really bad.